### PR TITLE
timeout

### PR DESCRIPTION
**Pull request by furynick**
Add timeout to pulsein as HC-SR04 max range is about 4m.

### CODE REVIEW

**Comment by Martinsos** [comment]
Hm interesting! So 30ms is the time it takes to measure 5m? And third argument to pulseIn is timeout after which what is returned -> the duration of timeout itself? That sounds good to me then! If what I wrote in previous sentence is correct, then I am ok with the change. I would only suggest removing comments and using variable instead -> we can use smth like
```
const int durationMicroSecFor5Meters = 30000
```

**Review by Martinsos — COMMENTED** [review]
I wrote a comment!

**Comment by furynick** [comment]
Agreed with variable usage, I'll add a commit to make a calculation to set maximum delay for a certain distance.

**Comment by furynick** [comment]
Commit pushed but needs some tests before merge.

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp:13]: I would not allow this for now, it will just complicate things. If people ask for it we can add it, but I doubt it will be important.
We could, however, create a constant maxDistanceCm, which can't be set from outside, and is always set to 400cm.

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp]: I would name the variable `maxDistanceDurationMicroSec` and then I would use it below as third argument like this `maxDistanceDurationMicroSec * 1.25`.
Also, shouldn't it be:
`maxDistanceDurationMicroSec = 2 * maxDistanceCm / speedOfSoundInCmPerMs`?

**Review by Martinsos — CHANGES_REQUESTED** [review]

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp:13]: Not really more complicated as the default value is set in constructor, both your examples are still running as-is without any modification.

Comment (Martinsos) [reply src/HCSR04.cpp:13]: True, but I was actually thinking about the case where users provide value bigger than 400cm. What then? We will need to explain that in that case, measurement will be incorrect. On the other hand, why would they provide value lesser than 400cm? To gain some performance? Ok, but is that really useful? I am not sure, that is why I thought, let's see, if people need it, we can add it.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp]: You're right, I forgot speedOfSoundInCmPerMs is already a double variable so no need to keep integer calulations but I think it's a bad idea to pass a floating point value to pulseIn.
In an ideal world I think maxDistanceDurationMicroSec need to be rounded and casted to ensure portability.

If I made no mistake speedOfSound unit is Cm per milliseconds, you have to multiply by 1000 to pass microseconds timeout parameter to [pulseIn](https://www.arduino.cc/reference/en/language/functions/advanced-io/pulsein/).

Saying that I have a doubt : 331.3 m/s = 0.3313 m/ms = 0.00033313 m/µs
Isn't there a mistake in caclulation of speedOfSoundInCmPerMs or I have missed something ?

I'll add a commit to rename variable as you wish.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp:13]: Moreover I think it's more versatile to use a variable as you can use two or more objects in same code for different sensors with different characteristics.
A global constant prevent usage of different sensors.

Comment (furynick) [reply src/HCSR04.cpp:13]: In case the sensor didn't respond your program is locked for 1 sec, if the distance needs to be checked every second and sensor is defective or distance can't be measured your code will be erratic.

To prevent locking code it must use a timeout in acceptable range.

File: src/HCSR04.cpp
Comment (Martinsos) [reply src/HCSR04.cpp]: Take a look at line 39: you will see `double distanceCm = durationMicroSec / 2.0 * speedOfSoundInCmPerMs;`. Assuming this works correctly, it means that `Ms` stands for microseconds! But it is very confusing I agree -> if we are dealing with microseconds, we should be explicit about it and certainly not use `ms` as abbreviation for it, since `ms` is normally used for milliseconds.
So, let's also rename `speedOfSoundInCmPerMs` to `speedOfSoundInCmPerMicroSec`, and let's not use `ms` anywhere but let's instead use `MicroSec`.

Regarding rounding -> makes sense! I am not sure what pulseIn takes, I will leave that to you, but I don't have anything against rounding `maxDistanceDurationMicroSec`, I don't think 1 microsecond will change much.

File: src/HCSR04.cpp
Comment (Martinsos) [reply src/HCSR04.cpp:13]: That does make sense! Well, maybe having this option with default set to 400cm, as you suggested, is ok. Being able to have some control over how long will the measuring action maximally take does sound important. But thinking further about it, maybe people would rather provide the timeout directly then, if it is important for them to ensure measuring operation does not take longer than specific time?
I can imagine couple use cases:
1. I know I don't care about distances larger than N cm. In that case, it makes sense to set max distance to N, so each measurement takes less time, and I can do more of them in a period of time, so I get higher resolution.
2. I know that sensor does not support distance larger than N cm.
3. I want to ensure that measurement does not take longer than T time, because I have some requirements in my real time system.

Use cases 1 and 2 are well supported by the change you suggested. There is a tricky edge case -> what if thy specify max distance larger than what their sensor supports? They might be getting weird results. But I guess this is on them. We can mention in documentation that default value provided is for the hc-sr04 sensor and if they are providing different value, they should ensure it does not go beyond maximum of what their sensor can do.

Use case 3 -> well, not really, because they would have to calculate the distance to achieve that time, and the calculation we do to get time based on distance is an "implementation detail" -> we don't want them to bother with that. So if we want to support this use case, we should provide them with a direct way to specify max time. Maybe we could offer them an additional setting, maxMeasurementDurationInMicroSec? Not sure if it should be constructor argument or argument of measure() function. And then we would take use min(maxMeasurementDurationInMicroSec, durationMicroSecForMaxDistance) and use that as a timeout.

Well this is now getting complicated, which is why I initially didn't want to get into it, but your reasons sound good, and if we are doing it let's go all the way! What do you think?

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp]: You're definitely right, I forgot m to cm conversion.

Renaming ok and max distance calulation fixed (inverted units).

I made some tests with sensors and everything is working.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp:13]: For use cases 1 & 2 weird results are impossible, if maxDistance is greater than sensor abilities, sensor will not echo before pulseIn timeout so function will returns -1.

Agreed for use case 3, let me dig into that to find the proper way to do it.

Comment (furynick) [reply src/HCSR04.cpp:13]: I don't think usage of min() is a good idea, if one decide to set a fixed value for timeout we shouldn't alter this value.

I guess using a parameter of the measure method will be hard to maintain if there are several call to this method on the code.

Perhaps we can add a method to specify an absolute timeout, default property is 0 which let automatic value within measure otherwise the fixed value is used.

Do you agree ?

Comment (Martinsos) [reply src/HCSR04.cpp:13]: @furynick, so sorry for dropping this! I am back if you want to finalize it. Let's see -> so you are proposing we allow them to set a timeout if they want, and in that case, that timeout is respected, not the maxDistance. And it should be an instance method (maybe contrcutors parameter?) -> ok that sounds good.

I am still in favor of using min() though. If we have maxDistance and maxTimeout, then it is clear that they are both max -> it means it might be less! But if I set maxDistance and then I set timeout and suddenly that distance is not respected any more -> well that sounds like something that is unexpected.
What do you say? So I am voting to add maxTimeout next to maxDistance, they are both equivalent in the sense that min() on both of them is used to figure out the actual timeout. And of course, maxTimeout would by default have negative value which we then handle in a special way.

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp]: Instead of this being a method, could we make it a parameter of constructor?
I prefer defining objects once and having them be immutable, then being able to change their behavior on the fly, unless it is really needed for a specific use case.

Comment (furynick) [reply src/HCSR04.cpp]: Ok for me, commit pushed.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp:13]: New commit pushed, for adding min between maxdistance & timeout, it might be a good idea.
Timeout is set to 0 by default (disabled).

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp]: Can we rename timeoutMicroSec to maxTimeoutMicroSec? That way it is clear that it might be shorter (due to maxDistanceCm).

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp]: Hm well this doesn't do anything, expression is not assigned!

How about we do
```
uint32_t maxDistanceDurationMicroSec = 2.5 * maxDistanceCm / speedOfSoundInCmPerMicroSec;
uint32_t timeoutMicroSec = min(maxDistanceDurationMicroSec, maxTimeoutMicroSec);

unsigned long durationMicroSec = pulseIn(echoPin, HIGH, timeoutMicroSec);

```

Notice I moved the declaration of maxDistanceDurationMicroSec from above diretly to the defintion -> we arein cpp so we can do this.

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp:43]: This is returning unsigned long -> should we maybe also use unsigned long for microsec then, instead of uint32_t?

File: src/HCSR04.h
Comment (Martinsos) [inline src/HCSR04.h:16]: We could offer more information here -> we could say that it is advisable not to set this bigger than what specification of the sensor says it can do (which is 4m for HC-SR04). And we could say that it can be useful to set it to smaller value if we don't care about distances above that because the measurement will take proportionally less time.

File: src/HCSR04.h
Comment (Martinsos) [inline src/HCSR04.h]: We could mention here that this is useful if you need to be sure that measurement will not take longer than some specific time. But, this also means that we are limiting max distance it can measure.

File: src/HCSR04.h
Comment (Martinsos) [inline src/HCSR04.h]: I would mention here that it takes into consideration maxDistanceCm and maxTimeoutMicroSec and takes both into account.

**Review by Martinsos — CHANGES_REQUESTED** [review]
Great, I think we are almost there! Left some more comments.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp:43]: unsigned long is equivalent to uint32_t but this type is clearer for devs so I changed it to uint32_t to keep every variables to be of the same type.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp]: you're right, fixed

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp]: "or use absolute timeout" -> that is not true any more, not we do minimum of those two. Maybe it is best to omit this last part, it is visible from the code.

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp:39]: I would not mutate `maxDistanceDurationMicroSec`, it has its meaning and here we set it to something else.
I suggest doing `const uint32_t maxDistanceDurationMicroSec = ...` above (so combining declaration with definition and also making it const and for the result of minimum we create new variable. You can use ternary operator to do if-else logic if you wish, that could be nice.

File: src/HCSR04.h
Comment (Martinsos) [inline src/HCSR04.h]: Update the var name.

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp]: I did short investigation in uint32_t and I am slightly concerned with using it and unsigned long interchangeably. This works because it seems that on arduino unsigned long actually is uint32_t, but if we don't need to be making this assumption, why do it? Why not just use unsigned long all the way and we are good?

**Review by Martinsos — COMMENTED** [review]
Left more comments!

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp:39]: uint32_t or unsigned long uses 4 bytes of memory, this is intended for microcontrollers not for PC with GB of RAM, it's not necessary to consume memory just to be more readable.
I don't think most peoples read libraries code, they just use it.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp]: I think you're right, it's more convenient to use the same type as as pulseIn.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp:39]: I would use float instead of double, sensors precision is about milimeters when they are excellent, double consumes 8 bytes.
float have 6 decimals that represents 10nm and double have about 14 decimals where we need at very most 2 !

Comment (Martinsos) [reply src/HCSR04.cpp:39]: @furynick hm I haven't thought about that (the memory)! I am not suggesting this for the people using the library btw., is for us maintainers to have clean code and avoid future mistakes.
The memory -> do we really have to think about it like that? I would expect compiler to optimize the usage of memory in cases like this. Is this a common practice, to try to not create new variables when working with Arduino/embedded code? What is your experience with it?

Float instead of double -> where are we using float? We are using integers, right?

Comment (Martinsos) [reply src/HCSR04.cpp:39]: Small advice -> don't resolve comments right after you answer them, otherwise I might miss them. In my experience resolve is most useful when used to acknowledge that last comment has been read and the person that has not written it (in this case me) has nothing to add.

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp:19]: Aha ok, this is where we return float instead of double!
I think you are right about returning float instead of double, that sounds good.
But in that case let's update the version of library to 2.0.0 since this is a breaking change.

File: src/HCSR04.cpp
Comment (Martinsos) [inline src/HCSR04.cpp]: While at it, should we also make maxDistanceCm into unsigned int instead of just int? We could even make it unsigned short I guess if we wanted to preserve some memory?

**Review by Martinsos — COMMENTED** [review]
Ok, I left some more comments but it looks good now, let's just finalize those. Reusing the variable when doing min() -> ok, if you think that makes a lot of sense on the optimization side, let's do it.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp:39]: Thanks for advice, I used them to mark comment resolved after committing related updates.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp:19]: Sure, that's a good idea.

File: src/HCSR04.cpp
Comment (furynick) [reply src/HCSR04.cpp]: I'll go forward and also change int to byte for pins, no change in memory on 8-bit AVR but int are 4-bytes long on 32-bit architectures and none of them have more than 256 digital pins (for now) as far as I know.

Comment (Martinsos) [reply src/HCSR04.cpp]: Ok, we are going wild with types :D! Sounds good, if it is not enough somebody will complain.

File: src/HCSR04.h
Comment (Martinsos) [inline src/HCSR04.h]: exceed

**Review by Martinsos — APPROVED** [review]
All right, looks good to me! Left one type comment. I haven't run any tests -> do you mind running some tests for the very end and then I will merge it?

**Comment by Martinsos** [comment]
@furynick Oh I actually forgot -> we also need to update the code example in README! Could you also update that one before we merge?

**Comment by furynick** [comment]
Done, hoping my english is not too bad ;)

File: src/HCSR04.h
Comment (Martinsos) [inline src/HCSR04.h]: You might want to set this value if you are using different sensor than HC-SR04 or if you don't care about distances larger than whatever you will set it to (therefore reducing time it takes for a single measurement).
```suggestion
     *                       You might want to set this value if you are using different sensor than HC-SR04
     *                       or if you don't care about distances larger than whatever you will set it to
     *                       (therefore reducing time it takes for a single measurement).
```

File: src/HCSR04.h
Comment (Martinsos) [inline src/HCSR04.h]:
```suggestion
     * @param maxTimeoutMicroSec  Single measurement will never take longer than whatever value you provide here.
     *   You might want to do this in order to ensure your program is never blocked for longer than some specific time,
     *   since measurements are blocking.
     *   By default, there is no limit on time (only on distance). By defining timeout, you are again limiting the distance. 
```

File: README.md
Comment (Martinsos) [inline README.md]:
```suggestion
Sensor is initialized by creating instance of class UltraSonicDistanceSensor and providing trigger and echo pins.
```

File: README.md
Comment (Martinsos) [inline README.md]:
```suggestion
Default value for maximum measurement distance is 4m, since HC-SR04 sensor can't measure reliably beyond that.
However, if you are using another sensor or if you you don't care about distances larger than some value, you can set a maximum distance in cm yourself.
```

File: README.md
Comment (Martinsos) [inline README.md]:
```suggestion
Besides defining max distance, you can also define max time of measurement (in micro seconds). This is important when you want to ensure your program does not block for longer than specific period of time.
```

**Comment by Martinsos** [comment]
Sorry @furynick for dropping this at some point -> I took some time today to polish it just a bit more and I merged it! Thanks for help with this :)!!

**Pull request merged**
